bundles/org.eclipse.text/src/org/eclipse/jface/text/source/AnnotationModel.java

```diff
@@ -639,11 +640,12 @@ private void cleanup(boolean fireModelChanged, boolean forkNotification) {
 		if (fDocumentChanged) {
 			fDocumentChanged= false;
 
-			ArrayList<Annotation> deleted= new ArrayList<>();
+			List<Annotation> deleted;
 			IAnnotationMap annotations= getAnnotationMap();
 			Object mapLock = annotations.getLockObject();
 			
 			if (mapLock == null) {
+				deleted= new ArrayList<>();
```

Comment:
Why not using streams here as well? It seems the only difference is the lock here?

Reply:
If the lock is not available, we might run into concurrent modifications. These are prevented by the defensive copy behind `keySetIterator`. It's a safety measure to not break existing subtypes / weird usages of the AnnotationModel.

Reply:
You add a default method to `IAnnotationMap` like this:

```
default Stream<Annotation> keyStream() {
	return StreamSupport.stream(
			Spliterators.spliteratorUnknownSize(keySetIterator(), Spliterator.ORDERED),
			false);
}
```

then it could be used as a stream as well.

Reply:
I thought default methods in interfaces were forbidden [because of incompatibility issues](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.pde.doc.user%2Freference%2Fapi-tooling%2Fapi_evolution.htm)? Or are they just discouraged?

In any case I tried wit a spliterator and the performance dropped drastically.

Reply:
It might be worth to record then if cleanup is really required, e.g setting a boolean if something is removed/cleared in the model an check this first then cleanup will only happen if there are likely item to be cleaned.

Reply:
> It might be worth to record then if cleanup is really required, e.g setting a boolean if something is removed/cleared in the model an check this first then cleanup will only happen if there are likely item to be cleaned.

I think that's what `fDocumentChanged` is supposed to do, the problem is that `FindReplaceDialog::replaceAll` finds each occurrence (which ends up calling `getAnnotationsIterator` and `cleanup`), marks it and replaces it (which changes the document). This is the piece of code where that happens:

```java
int index= 0;
while (index != -1) {
	index = findAndSelect(...); // this one calls getAnnotationIterator and cleanup
	if (index != -1) { // substring not contained from current position
		// this one sets fDocumentChanged to true
		Point selection= replaceSelection(replaceString, regExSearch);
		...
	}
}
```

This is what you meant before, right @szarnekow ?

Reply:
I've been looking at it and I'm out of ideas: any suggestions on how to wrap all changes into one session and fire a document change at the end of the loop? Maybe we can discuss this in another PR?

Reply:
> Maybe we can discuss this in another PR?

On another **issue** please.

Reply:
I created https://github.com/eclipse-platform/eclipse.platform.ui/issues/995, feel free to take over.

Thank you all for your input and for helping improve this code even more!

I think this PR is now ready to be merged, would you agree?

Reply:
Without this change, I get fairly consistent numbers (showing only one of many experiments here):
```
ReplaceAll: 13543ms
Undo: 6926ms
Redo: 4773ms
Undo: 6311ms
ReplaceAll: 12204ms
Undo: 6989ms
Redo: 5155ms
Undo: 6349ms
ReplaceAll: 13077ms
```

After having applied this change locally, I see way more flaky numbers, though some of them are sometimes better than originally:
```
ReplaceAll: 11329ms
Undo: 8183ms
Redo: 4729ms
Undo: 6209ms
ReplaceAll: 9549ms
Undo: 6339ms
Redo: 4717ms
Undo: 6645ms
ReplaceAll: 23934ms.  <--- ???
Undo: 7332ms
ReplaceAll: 10243ms
Undo: 7099ms
ReplaceAll: 12019ms.
Undo: 7451ms
Redo: 5542ms
Undo: 6634ms
ReplaceAll: 9602ms
```

When I only look at the ReplaceAll timings, it's a nice improvement on avg. But the fact that the numbers seem to indicate degradation of `Undo` timings by some small percentage makes me sad. Mostly because the code *looks* fine, and I cannot explain these numbers yet. There might be an impact on other operations beyond the usage pattern behind ReplaceAll. 

What concerns me, too, is that I get consistently higher memory usage with the change applied - which I cannot explain now.

@iloveeclipse What do you think?

Reply:
I would assume depending on common pool state the times may vary. 
Considering there is not much improvement given, I would prefer to keep the old/simple/non parallel code and spend time on analysis / fix of #995.

Reply:
For completeness:
The numbers in the JavaEditor with Code-Folding disabled:
```
ReplaceAll: 2545ms
Undo: 4074ms
Redo: 2260ms
Undo: 4566ms
ReplaceAll: 4924ms
Undo: 4303ms
ReplaceAll: 2260ms
```

Same sequence of operations in the plain text editor:
```
ReplaceAll: 331ms
Undo: 188ms
Redo: 103ms
Undo: 141ms
ReplaceAll: 216ms
Undo: 252ms
ReplaceAll: 187ms
```